core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,221 @@
+  private final Path keystorePath;
+  private final Optional<SlashingProtection> slashingProtection;
+  private final ArtifactSignerProvider artifactSignerProvider;
+
+  public ImportKeystoresHandler(
+      final ObjectMapper objectMapper,
+      final Path keystorePath,
+      final Optional<SlashingProtection> slashingProtection,
+      final ArtifactSignerProvider artifactSignerProvider) {
+    this.objectMapper = objectMapper;
+    this.keystorePath = keystorePath;
+    this.slashingProtection = slashingProtection;
+    this.artifactSignerProvider = artifactSignerProvider;
+  }
+
+  @Override
+  public void handle(RoutingContext context) {
+    // API spec - https://github.com/ethereum/keymanager-APIs/tree/master/flows#import
+    final RequestParameters params = context.get("parsedParameters");
+    final ImportKeystoresRequestBody parsedBody;
+    try {
+      parsedBody = parseRequestBody(params);
+    } catch (final IllegalArgumentException | JsonProcessingException e) {
+      handleInvalidRequest(context, e);
+      return;
+    }
+
+    // check that keystores have matching passwords
+    if (parsedBody.getKeystores().size() != parsedBody.getPasswords().size()) {
+      context.fail(BAD_REQUEST);
+      return;
+    }
+
+    // no keystores passed in, nothing to do, return 200 and empty response.
+    if (parsedBody.getKeystores().isEmpty()) {
+      try {
+        context
+            .response()
+            .putHeader(CONTENT_TYPE, JSON_UTF_8)
+            .setStatusCode(SUCCESS)
+            .end(
+                objectMapper.writeValueAsString(
+                    new ImportKeystoresResponse(Collections.emptyList())));
+      } catch (JsonProcessingException e) {
+        context.fail(SERVER_ERROR, e);
+      }
+      return;
+    }
+
+    final Set<String> existingPubkeys = artifactSignerProvider.availableIdentifiers();
+    final List<ImportKeystoreResult> results = new ArrayList<>();
+    final List<String> importedPubkeys = new ArrayList<>();
+    for (int i = 0; i < parsedBody.getKeystores().size(); i++) {
+      try {
+        final String jsonKeystoreData = parsedBody.getKeystores().get(i);
+        final String password = parsedBody.getPasswords().get(i);
+        String pubkey = new JsonObject(jsonKeystoreData).getString("pubkey");
+        // the BLS keystore generator generates a pubkey without `0x` in front of it, breaks
+        // comparison
+        if (!pubkey.startsWith("0x")) {
+          pubkey = "0x" + pubkey;
+        }
+
+        if (existingPubkeys.contains(pubkey)) {
+          results.add(
+              new ImportKeystoreResult(ImportKeystoreStatus.DUPLICATE, "Pubkey already imported"));
+        } else {
+          createKeyStoreYamlFileAt(pubkey, jsonKeystoreData, password, KeyType.BLS);
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.IMPORTED, "Success"));
+          importedPubkeys.add(pubkey);
+        }
+      } catch (Exception e) {
+        results.add(
```

Comment:
Need to cleanup and remove the failed individual keystore as well

Reply:
done